### PR TITLE
Initial Python3 support with Test-Driven-Development enforced by CI (with git diff coverage and diff lint checking) + mypy, pyre and pytype

### PR DESCRIPTION
**Pull request by bernhardkaindl**
I have worked on @ydirson's PR #17 and think that my work on it should be ready now.

With this PR, Python3 support is not 100% complete yet, but with all tests passing for Python3 and all non-whitespace changes covered by code coverage, it's a big leap forward.

Warnings and further bugs remain, but they have to be fixed in my other PRs!

The summary of reviews done so far is: It's fine, and I shall update the CodeCov upload to upload code coverage for Python2.7 and Python3 separately, but I can do this in a followup PR, no problem for now:

### Initial Python3 support with Test-Driven-Development enforced by CI:

Initial Python3 support with Test-Driven-Development enforced by CI:
- with git diff coverage and diff lint checking
- with Codecov upload
- with Static analysis using `mypy`, `pyre` and `pytype`

GitHub CI (but also local CI checks) that any changes (besides white space-only diff lines):
- have 100% code coverage of the diff to origin/master
- do not introduce a `pylint` warning type which is not disabled in `pylintrc`
- do not introduce a new type of static analysis warning (as long as it is not currently suppressed) in `mypy`, `pyre`, and `pytype`.

Documentation on setting up local CI using tox and more is in [README.md](https://github.com/xenserver-next/python-libs/blob/py3-move-to-tox/README.md).

Follow-up: 
- The warnings shown (in the GitHub Actions Summary Page) are reminders that the PRs 22, 23, 24 are needed, continue with them.

### CODE REVIEW

File: requirements-dev.txt [hunk 1]
```diff
@@ -1,10 +1,28 @@
-# necessary dev/test tasks
-pylint
-coverage
-diff_cover
+# Used by tox jobs to install test suite dependencies.
+# CI dependencies are only required in the specific tox environments.
+# Maybe move these as development deps to pyproject.toml later:
```
Comment (psafont) [inline requirements-dev.txt:3]: These can be done with:

```toml
[project.optional-dependencies]
test = [
    "mock",
    "pytest",
    "pytest-cov",
    "typing_extensions"
]
```
Then on tox.ini:
```
-deps        = -r requirements-dev.txt
+extras      = test
```
See [extras](https://tox.wiki/en/4.5.1/config.html#extras)

After that, the all requirements*.txt should be able to be deleted

Comment (bernhardkaindl) [reply requirements-dev.txt]: Done!
- Works nicely with Python3.7 which supports tox version 4
- Needed to use Python3.8 to run tox version 4 and specify to run the 3.6 and 2.7 tests.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,192 @@
+[tox]
+envlist = py311-pyre, py310-pytype, py39-check, py38-covcombine, py37-mdreport, py36-lint
+isolated_build = true
+skip_missing_interpreters = true
+
+# TODO later: Enable the venv_update extension so that changes to requirements-dev.txt
+# are automatically detected.
+tox_pip_extensions_ext_venv_update = true
```
Comment (psafont) [inline tox.ini:8]: This looks like it can be deleted

**Review by psafont — APPROVED** [review]

File: xcp/net/mac.py [hunk 3]
```diff
@@ -89,7 +89,7 @@ def _set_from_str_octets(self, octets):
 
         self.octets = [ int(i, 16) for i in octets ]
         self.integer = sum(t[0] << t[1] for t in
-                           zip(self.octets, range(40, -1, -8)))
+                           zip(self.octets, list(range(40, -1, -8))))
```
[the remark this answers is not in the document]
Comment (bernhardkaindl) [reply xcp/net/mac.py]: Correct, thanks! It was a spurious false positive from `python2 -m pylint --py3k` which
is explained for 2to3 here: https://diveintopython3.net/porting-code-to-python-3-with-2to3.html#xrange
> If your Python 2 code used range(), the 2to3 script does not know whether you needed a list, or whether an iterator would do. It errs on the side of caution and coerces the return value into a list by calling the list() function.

Indeed, as you said:
> zip() expects an iterable and the result of range() is iterable.

Replaced the superflous 2to3 change with a suppression of this pylint warning!

**Review by rosslagerwall — APPROVED** [review]

**Pull request merged**